### ISSUE TITLE
"Uncaught TypeError: Cannot read property 'length' of undefined" after upgrading

### ISSUE THREAD

**Issue by fionera** [ether/etherpad-lite#4208]
**Describe the bug**
Like #2107 after migrating to 1.8.4 some pads dont work and throw the following error:
```
ErrorId: GERrFJa8w9CcmYQRig1h
URL: https://pad.ctdo.de/p/einkaufsliste
UserAgent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36
Uncaught TypeError: Cannot read property 'length' of undefined in https://pad.ctdo.de/javascripts/lib/ep_etherpad-lite/static/js/timeslider.js?callback=require.define&v=e698cbe0 at line 22
```

**To Reproduce**
1. Use 1.6.3
2. Upgrade to 1.8.4
3. See error

**Expected behavior**
The pads should work

**Screenshots**
-

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser: Chrome
 - Version: 83

**Additional context**
#2107

**Comment by JohnMcLear** [comment]
Export the .Etherpad of the pad.

/p/whatever/export/etherpad

And share it somewhere someone can access...

**Comment by fionera** [comment]
Hi :) Thanks for the fast response. Here the exported File: https://trash.ctdo.de/b/5f2494c1d04/einkaufsliste.etherpad

**Comment by JohnMcLear** [comment]
```

[2020-08-02 12:20:46.616] [WARN] message - Dropped message, USERINFO_UPDATE Color is malformed.[object Object]
[2020-08-02 12:20:54.015] [INFO] access - [LEAVE] Pad "7wlg-jdivkXBgQhDKknX": Author "a.Vh6RJ1wVlqGLFs4T" on client 9dCHKiis9tD5lXYoAAAA with IP "192.168.77.177" left the pad
[2020-08-02 12:20:54.738] [INFO] access - [ENTER] Pad "7wlg-jdivkXBgQhDKknX": Client tULoV3L91qeSIdPGAAAB with IP "192.168.77.177" entered the pad
[2020-08-02 12:20:55.190] [WARN] message - Dropped message, USERINFO_UPDATE Color is malformed.[object Object]

```

on client

```
Uncaught Error: Failed assertion: Invalid changeset (checkRep failed)
    error http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:5824
    assert http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:5838
    checkRep http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:6069
    setDocAText http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:716
    setDocumentAttributedText http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:184
    setBaseAttributedText http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:10257
    withCallbacks http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:180
    inCallStack http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:437
    inCallStackIfNecessary http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:503
    withCallbacks http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:177
    setBaseAttributedText http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:10243
    ace_setBaseAttributedText http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:944
    Ace2Editor/</editor[fnName]< http://192.168.77.180:9001/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define&v=351053b7:2913
    action http://192.168.77.180:9001/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define&v=351053b7:2866
    doActionsPendingInit http://192.168.77.180:9001/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define&v=351053b7:2886
    forEach http://192.168.77.180:9001/javascripts/lib/ep_etherpad-lite/static/js/ace2_common.js?callback=require.define&v=351053b7:10613
    doActionsPendingInit http://192.168.77.180:9001/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define&v=351053b7:2885
    onEditorReady http://192.168.77.180:9001/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define&v=351053b7:3018
    readyFunc http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 3153 > scriptElement:17
    init http://192.168.77.180:9001/p/7VIY7Wakte9rAYay7DcY?noColors=false&showControls=true&showChat=true&showLineNumbers=true&useMonospaceFont=false&userName=false&userColor=false&rtl=false&alwaysShowChat=false&chatAndUsers=false&lang=en-gb line 24 > scriptElement line 234 > Function:5363
7VIY7
```

I feel like something is missing here, like some information, for example..  

What database are you using?

Are you using plugins?

Afaik this is all requested in the issue template though so I can't imagine you omitted this intentionally so **I have assumed dirtydb and no plugins as is the default.**  Can you please confirm?

**Comment by fionera** [comment]
Oh sorry it was very late when I opened the issue. I am using mysql as database and no additonal plugins.

**Comment by JohnMcLear** [comment]
I think your database is misconfigured.  Check your charsets etc.

**Comment by fionera** [comment]
I use utf8mb4 as Charset in the ENV Variables and the Database has the following config:
```
MariaDB [(none)]> SELECT default_character_set_name FROM information_schema.SCHEMATA 
    -> WHERE schema_name = "ctdopad";
+----------------------------+
| default_character_set_name |
+----------------------------+
| utf8mb4                    |
+----------------------------+
```

```
MariaDB [(none)]> SELECT CCSA.character_set_name FROM information_schema.`TABLES` T,        information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHERE CCSA.collation_name = T.table_collation   AND T.table_schema = "ctdopad";
+--------------------+
| character_set_name |
+--------------------+
| utf8mb4            |
+--------------------+
```

Is there any way how I can check if the data is correct in the db, like a cleanup command to delete corrupt data?

**Comment by JohnMcLear** [comment]
Looks right to me.

There is checkPad.js but I don't know how reliable it is at the moment

**Comment by webzwo0i** [comment]
https://gist.github.com/webzwo0i/c6d75c09c4db0599ded42e7988cb39aa#file-checkpaddeltas-js this is an fast fix of checkPadDelta.js ,just calls checkRep on every revision and tries to applyToAText and comparing the result for every key revision

copy it to bin/ and call it like node bin/checkPadDeltas.js your-pad-name

**Comment by webzwo0i** [comment]
Can you also check your store table character set
`SELECT CCSA.character_set_name FROM information_schema.`TABLES` T,information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHERE CCSA.collation_name = T.table_collation AND T.table_schema = 'ctdopad' AND T.table_name = 'store'`
and can you paste the relevant lines of your my.cnf (especially init_connect if its there)
and also
`show variables like 'character_set%';`  I think relevant are character_set_client and character_set_connection

**Comment by mmccarn** [comment]
I was getting this error after updating from v1.8 to v1.8.6 - but only after 'waking' my computer to a pad that was left open when the system went to sleep.

My charset was still "utf8".  I have followed the instructions from [Consider MySQL UTF8 issues](https://github.com/ether/etherpad-lite/issues/2516) to convert my database and etherpad settings from utf8 to utf8mb4

System details:
* Centos6 
* mysql55 from RedHat Software Collections
* [.my.cnf](https://dev.mysql.com/doc/refman/5.7/en/option-files.html#option-files-unix) includes database login details, so "-u [username] -p" is not required in the commands below
* etherpad-lite is installed in /opt/etherpad-lite
* node v10.22.1
* etherpad-lite running as a supervised [runit](http://smarden.org/runit/) service (https://github.com/mmccarn/smeserver/tree/master/var/service/etherpad-lite)

```
Stop the service
# sv d etherpad-lite

Backup the database
# /opt/rh/mysql55/root/usr/bin/mysqldump --socket=/var/lib/mysql/mysql55.sock --databases etherpad |gzip > /ibays/Primary/html/etherpad/etherpad_db.sql.gz

# /opt/rh/mysql55/root/usr/bin/mysqlcheck --socket=/var/lib/mysql/mysql55.sock etherpad
etherpad.store
warning  : 104 clients are using or haven't closed the table properly
status   : OK

# mysql55 etherpad -e 'alter database `etherpad` CHARACTER SET = utf8mb4 COLLATE = utf8mb4_bin;'

# mysql55 etherpad -e 'alter table `store` convert to character set utf8mb4 collate utf8mb4_bin;'
# mysql55 etherpad -e 'alter table `store` change `key` `key` varchar(100) character set utf8mb4 collate utf8mb4_bin;'
# mysql55 etherpad -e 'alter table `store` change `value` `value` longtext character set utf8mb4 collate utf8mb4_bin;'

# sed -i 's/"utf8"/"utf8mb4"/g' /opt/etherpad-lite/settings.json

Restart the service
sv u etherpad-lite
```

**Comment by rhansen** [comment]
> but only after 'waking' my computer to a pad that was left open when the system went to sleep.

@mmccarn Can you try the current tip of the `develop` branch? There have been some relevant changes since v1.8.6 was released (see https://github.com/ether/etherpad-lite/issues/4331).

**Comment by mmccarn** [comment]
I think I'm running the ```develop``` branch now. At least, ```.../admin/plugins/info``` reports

Git sha: [127923e](https://github.com/ether/etherpad-lite/commit/127923e)

**Comment by ptandler** [comment]
We just had a small test run with the current develop version (well, the version from last Monday) and this happened several times. But not for all users, maybe 5%-10%. At least they got this message. And we did no upgrade, but used a fresh install.

**Comment by JohnMcLear** [comment]
@ptandler please provide some details.  Any plugins?  What settings ??  Etc etc.  See guidelines for new issues for the info we request on bug reports

**Comment by ptandler** [comment]
@JohnMcLear today we again did a load test of our etherpad server and this happened again.
But I looked at the details and I'm pretty sure that it is a different error, so I file a new issue: https://github.com/ether/etherpad-lite/issues/5047
